## Tarefa 01

- Professor: Fabrício Galende Marques de Carvalho
- Aluno: André Salerno (RA: 1461392411015)

---

TC.1. Exemplificar, seguindo as definições de frase, documento, corpus e corpora, cada um desses elementos considerando os seguintes contextos:

(a) Revisões de produtos em sites de e-commerce, tais como mercado livre e amazon.

(b) Solicitações enviadas à uma equipe de helpdesk.

(c) Livros técnicos

(d) Livros não técnicos (e.g. romances)

(e) Logs de interações com um chatbot.


---

PC.1. Desenvolva um programa que acesse e exiba os atributos dos corpora que são enumerados a seguir.
Adicionalmente, descreva as principais características, tais como origem do texto, metadados, ano
dos textos, anos de produção, etc. que fazem parte desses corpora. Utilize a biblioteca NLTK.

(a) Machado.

(b) Mac-Morpho.

(c) Floresta Sintáctica

---

## TC.1 — Frase, documento, corpus e corpora em diferentes contextos

### Definições adotadas

<table>
<thead><tr><th>Elemento</th><th>Definição</th></tr></thead>
<tbody>
<tr><td><strong>Frase (sentença)</strong></td><td>Menor unidade textual que expressa uma ideia relativamente completa; sequência de palavras normalmente delimitada por pontuação forte (<code>.</code>, <code>?</code>, <code>!</code>) ou por quebra de turno. É a unidade típica de tarefas como <em>tokenization</em>, <em>parsing</em>, tradução e análise de sentimento em nível de sentença.</td></tr>
<tr><td><strong>Documento</strong></td><td>Unidade de texto coesa e autocontida, produzida em <strong>um único contexto comunicativo</strong> (um autor, um evento, uma interação). É formado por uma ou mais frases e costuma carregar <strong>metadados próprios</strong> (autor, data, título, categoria, avaliação…).</td></tr>
<tr><td><strong>Corpus</strong></td><td>Coleção <strong>estruturada e finita</strong> de documentos, reunidos segundo um <strong>critério explícito</strong> (gênero, época, fonte, língua, domínio). Em geral acompanha metadados e, às vezes, anotações linguísticas (POS, árvores sintáticas). Serve de base para estudo linguístico ou para treino/avaliação de modelos de PLN.</td></tr>
<tr><td><strong>Corpora</strong></td><td>Plural de <em>corpus</em>: <strong>conjunto de dois ou mais corpora</strong> tratados em conjunto — frequentemente de fontes, gêneros ou línguas diferentes — para estudos comparativos ou para ampliar a cobertura de um modelo.</td></tr>
</tbody>
</table>

> Relação hierárquica: **palavra ⊂ frase ⊂ documento ⊂ corpus ⊂ corpora**.

---

### (a) Revisões de produtos em e-commerce (Mercado Livre, Amazon)

<table>
<thead><tr><th>Elemento</th><th>Exemplo</th></tr></thead>
<tbody>
<tr><td><strong>Frase</strong></td><td>"O produto chegou dois dias antes do prazo e veio muito bem embalado."</td></tr>
<tr><td><strong>Documento</strong></td><td>Uma avaliação completa de um usuário sobre um item: nota (1–5 estrelas), título, texto livre, data, marcação "compra verificada" e contagem de "útil".</td></tr>
<tr><td><strong>Corpus</strong></td><td>Todas as avaliações da categoria <em>Fones de ouvido</em> da Amazon Brasil coletadas em 2023 (≈ 80 mil avaliações), com os metadados nota, data, verificado e votos de utilidade.</td></tr>
<tr><td><strong>Corpora</strong></td><td>União do corpus de avaliações da Amazon + corpus de avaliações do Mercado Livre + corpus de avaliações da Americanas, reunidos para um estudo comparativo de análise de sentimento entre plataformas.</td></tr>
</tbody>
</table>

### (b) Solicitações enviadas a uma equipe de helpdesk

<table>
<thead><tr><th>Elemento</th><th>Exemplo</th></tr></thead>
<tbody>
<tr><td><strong>Frase</strong></td><td>"Não consigo acessar minha conta corporativa desde ontem à tarde."</td></tr>
<tr><td><strong>Documento</strong></td><td>Um chamado (ticket) completo: assunto, descrição do problema, categoria, prioridade, solicitante e todo o histórico de mensagens entre usuário e atendente até o fechamento.</td></tr>
<tr><td><strong>Corpus</strong></td><td>Todos os tickets abertos no sistema de helpdesk de TI da empresa durante o ano de 2024, com os metadados categoria, prioridade, tempo de resolução, cumprimento de SLA e nota de satisfação.</td></tr>
<tr><td><strong>Corpora</strong></td><td>Tickets do helpdesk de TI + tickets do helpdesk de RH + tickets do suporte ao cliente externo, reunidos para treinar um classificador que decide para qual fila encaminhar cada nova solicitação.</td></tr>
</tbody>
</table>

### (c) Livros técnicos

<table>
<thead><tr><th>Elemento</th><th>Exemplo</th></tr></thead>
<tbody>
<tr><td><strong>Frase</strong></td><td>"A complexidade de tempo do algoritmo de busca binária é O(log n) no pior caso."</td></tr>
<tr><td><strong>Documento</strong></td><td>Um livro técnico inteiro (ou um capítulo isolado), p.ex. <em>Introduction to Algorithms</em> (Cormen et al.), com metadados autor(es), editora, ano, edição e ISBN.</td></tr>
<tr><td><strong>Corpus</strong></td><td>Uma coleção de 500 livros didáticos de Ciência da Computação publicados entre 2000 e 2020 por uma editora acadêmica, organizados por subárea (algoritmos, redes, banco de dados…).</td></tr>
<tr><td><strong>Corpora</strong></td><td>Corpus de livros técnicos de Computação + corpus de livros de Engenharia Elétrica + corpus de livros de Medicina, reunidos para um estudo de extração automática de terminologia científica.</td></tr>
</tbody>
</table>

### (d) Livros não técnicos (romances)

<table>
<thead><tr><th>Elemento</th><th>Exemplo</th></tr></thead>
<tbody>
<tr><td><strong>Frase</strong></td><td>"Dona Plácida ainda viveu alguns anos nesse estado." (<em>Memórias Póstumas de Brás Cubas</em>)</td></tr>
<tr><td><strong>Documento</strong></td><td>Um romance completo, p.ex. <em>Dom Casmurro</em> (1899), com metadados autor, ano de publicação, gênero e editora.</td></tr>
<tr><td><strong>Corpus</strong></td><td>A obra romanesca completa de Machado de Assis — os 10 romances publicados entre 1872 e 1908 (exatamente o que a categoria <code>romance</code> do corpus <strong>Machado</strong> do NLTK contém).</td></tr>
<tr><td><strong>Corpora</strong></td><td>Romances de Machado de Assis + romances de José de Alencar + romances de Eça de Queirós, reunidos para um estudo do romance realista de língua portuguesa no século XIX.</td></tr>
</tbody>
</table>

### (e) Logs de interações com um chatbot

<table>
<thead><tr><th>Elemento</th><th>Exemplo</th></tr></thead>
<tbody>
<tr><td><strong>Frase</strong></td><td>Turno do usuário: "Qual é o horário de funcionamento da loja?" — Turno do bot: "A loja funciona de segunda a sábado, das 9h às 18h."</td></tr>
<tr><td><strong>Documento</strong></td><td>Uma sessão de conversa completa (do primeiro ao último turno), com todos os pares usuário↔bot, <em>timestamps</em>, <em>intent</em> detectado em cada turno, confiança do classificador e identificador da sessão.</td></tr>
<tr><td><strong>Corpus</strong></td><td>Todos os logs de conversa do chatbot de atendimento coletados em um mês, com os metadados <em>intent</em>, confiança, ocorrência de <em>fallback</em>, canal e avaliação final do usuário.</td></tr>
<tr><td><strong>Corpora</strong></td><td>Logs do chatbot do site + logs do bot do WhatsApp + logs do assistente do aplicativo, reunidos para analisar a cobertura de <em>intents</em> e a consistência das respostas entre canais.</td></tr>
</tbody>
</table>


---

## PC.1 — Atributos dos corpora `machado`, `mac_morpho` e `floresta` (NLTK)

O programa abaixo:

1. baixa os três corpora com `nltk.download`;
2. exibe os **atributos comuns** de cada *CorpusReader* (classe, diretório raiz, codificação, número de arquivos, `fileids`, métodos de acesso disponíveis);
3. exibe **atributos específicos** de cada um (categorias, contagem de palavras/frases/parágrafos, conjunto de etiquetas, `README` e amostras de conteúdo);
4. é seguido de uma célula de texto descrevendo **origem, metadados, anos dos textos e ano de produção** de cada corpus.

> **Nota técnica:** na versão 3.9.1 do NLTK o *loader* `nltk.corpus.machado` está com um *bug* (o `PortugueseCategorizedPlaintextCorpusReader` não inicializa a superclasse, gerando `AttributeError: _fileids`). Por isso o corpus Machado é reconstruído manualmente com `CategorizedPlaintextCorpusReader` a partir dos mesmos arquivos baixados — o resultado é idêntico ao do *reader* oficial.


In [1]:
import nltk
from nltk.corpus.reader import CategorizedPlaintextCorpusReader

# 1) Garantir os corpora no disco ------------------------------------------------
# IMPORTANTE: no NLTK 3.9.1, chamar nltk.download() para um corpus que JA foi
# carregado em memoria corrompe o ponteiro interno do reader (_root perde o
# atributo _path) e qualquer acesso a `corpus.root` passa a lancar
#   AttributeError: 'FileSystemPathPointer' object has no attribute '_path'
# Por isso so baixamos o que ainda nao existe -- assim, ao reexecutar as
# celulas, nltk.download() nunca e chamado de novo.
# (Se o kernel ja estiver nesse estado, reinicie-o e rode tudo de novo.)
def garantir(pacote, recurso):
    try:
        nltk.data.find(recurso)
    except LookupError:
        nltk.download(pacote, quiet=True)

garantir("machado", "corpora/machado")
garantir("mac_morpho", "corpora/mac_morpho")
garantir("floresta", "corpora/floresta")

from nltk.corpus import mac_morpho, floresta   # noqa: E402  (apos garantir o download)


def _raiz(corpus):
    """str do diretorio raiz do corpus, resistente ao bug do _path no NLTK 3.9.1."""
    try:
        return str(corpus.root)
    except AttributeError:
        return "(indisponivel - reinicie o kernel e rode tudo de novo)"


def resumo_generico(nome, corpus):
    """Exibe, de forma compacta, os atributos comuns a qualquer CorpusReader."""
    fids = corpus.fileids()        # dispara o carregamento "preguiçoso" do reader
    try:
        enc = corpus.encoding(fids[0])
    except Exception:
        enc = "n/d"
    metodos = [m for m in ("raw", "words", "sents", "paras", "tagged_words",
                           "tagged_sents", "parsed_sents", "categories")
               if hasattr(corpus, m)]
    print(f"### {nome}  (reader: {type(corpus).__name__})")
    print("raiz     :", _raiz(corpus))
    print("cod./arq.:", enc, "/", len(fids), "arquivo(s)  ->", fids[:4],
          "..." if len(fids) > 4 else "")
    print("metodos  :", metodos)


In [2]:
# =========================================================================
# (a) MACHADO  --  Obra Completa de Machado de Assis
# =========================================================================
# Reconstrucao manual do reader (contorna o bug do loader oficial no NLTK 3.9.1)
raiz_machado = nltk.data.find("corpora/machado.zip/machado/")
machado = CategorizedPlaintextCorpusReader(
    raiz_machado, r"(?!\.).*\.txt",
    cat_pattern=r"([a-z]+)/.*",    # categoria = subdiretorio (romance, contos, ...)
    encoding="latin-1",
)

resumo_generico("(a) MACHADO", machado)

# --- Atributos especificos ----------------------------------------------------
print("\ncategorias (generos):")
for cat in machado.categories():
    fids = machado.fileids(categories=cat)
    print(f"  {cat:11s} {len(fids):3d} arq.  ex.: {fids[0]}")
print("total de palavras no corpus:",
      sum(len(machado.words(f)) for f in machado.fileids()))

# README: catalogo com titulo + ANO de publicacao (secao dos romances)
readme_machado = nltk.data.find(
    "corpora/machado.zip/machado/README").open().read().decode("latin-1")
romances = readme_machado.split("Romance", 1)[1].split("Poesia", 1)[0].strip()
print("\n----- README: romances (titulo + ano) -----")
print(romances)

# Um DOCUMENTO do corpus: o romance "Dom Casmurro" (1899)
doc = "romance/marm08.txt"
print(f"\n----- Documento de exemplo: {doc} -----")
print(f"paragrafos={len(machado.paras(doc))}  frases={len(machado.sents(doc))}"
      f"  palavras={len(machado.words(doc))}")
print("metadados (inicio do arquivo):", " ".join(machado.words(doc)[:6]))
print("frase de exemplo            :", " ".join(machado.sents(doc)[6]))


### (a) MACHADO  (reader: CategorizedPlaintextCorpusReader)
raiz     : C:\Users\andre\AppData\Roaming\nltk_data\corpora\machado.zip\machado
cod./arq.: latin-1 / 246 arquivo(s)  -> ['contos/macn001.txt', 'contos/macn002.txt', 'contos/macn003.txt', 'contos/macn004.txt'] ...
metodos  : ['raw', 'words', 'sents', 'paras', 'categories']

categorias (generos):
  contos      137 arq.  ex.: contos/macn001.txt
  critica      45 arq.  ex.: critica/mact01.txt
  cronica      24 arq.  ex.: cronica/macr01.txt
  miscelanea   10 arq.  ex.: miscelanea/mams01.txt
  poesia        7 arq.  ex.: poesia/maps01.txt
  romance      10 arq.  ex.: romance/marm01.txt
  teatro       10 arq.  ex.: teatro/matt01.txt
  traducao      3 arq.  ex.: traducao/matr01.txt
total de palavras no corpus: 3121944

----- README: romances (titulo + ano) -----
romance/marm01.txt: Ressurreição (1872)
romance/marm02.txt: A Mão e a Luva (1874)
romance/marm03.txt: Helena (1876)
romance/marm04.txt: Iaiá Garcia (1878)
romance/marm05.txt: M

### (a) Machado — características

<table>
<thead><tr><th>Aspecto</th><th>Descrição</th></tr></thead>
<tbody>
<tr><td><strong>O que é</strong></td><td><em>Obra Completa</em> de Machado de Assis (1839–1908), o principal escritor do Realismo brasileiro. É um corpus <strong>literário monolíngue (português do Brasil), de autor único</strong>.</td></tr>
<tr><td><strong>Origem do texto</strong></td><td>Digitalização das <em>Obras Completas de Machado de Assis</em>, vol. I–III, editora Nova Aguilar (Rio de Janeiro, 1994), tendo como base as edições originais da Editora Garnier. Disponibilizada em <strong>domínio público</strong> pelo portal do Ministério da Educação (<code>machado.mec.gov.br</code>).</td></tr>
<tr><td><strong>Estrutura / metadados</strong></td><td>246 arquivos <code>.txt</code> organizados em 8 subdiretórios, que funcionam como <strong>categorias / gêneros</strong>: <code>romance</code>, <code>contos</code>, <code>cronica</code>, <code>critica</code>, <code>poesia</code>, <code>teatro</code>, <code>traducao</code> e <code>miscelanea</code>. O arquivo <code>README</code> (e <code>CONTENTS</code>) traz o <strong>catálogo completo com título e ano de cada obra</strong>; as primeiras linhas de cada <code>.txt</code> repetem gênero, título e ano.</td></tr>
<tr><td><strong>Anos dos textos</strong></td><td>Obras publicadas entre <strong>1854 e 1939</strong> (póstumas/dispersas). Exemplos: <em>Crisálidas</em> (1864), <em>Ressurreição</em> (1872), <em>Memórias Póstumas de Brás Cubas</em> (1881), <em>Quincas Borba</em> (1891), <em>Dom Casmurro</em> (1899), <em>Esaú e Jacó</em> (1904), <em>Memorial de Aires</em> (1908).</td></tr>
<tr><td><strong>Anotação</strong></td><td>Nenhuma — é <strong>texto puro</strong> (<em>plaintext</em>). O NLTK oferece acesso a <code>raw</code>, <code>words</code>, <code>sents</code>, <code>paras</code> e às <code>categories</code>.</td></tr>
<tr><td><strong>Ano de produção do corpus</strong></td><td>Compilação eletrônica feita pelo NILC/USP a partir da edição de 1994; incorporada ao NLTK Data (pacote <code>machado</code>, ~5 MB).</td></tr>
<tr><td><strong>Codificação / licença</strong></td><td>ISO-8859-1 (<code>latin-1</code>); domínio público.</td></tr>
<tr><td><strong>Uso típico</strong></td><td>Estudos de estilo e autoria, modelagem de linguagem histórica do português, geração de texto, análise literária computacional.</td></tr>
</tbody>
</table>


In [3]:
# =========================================================================
# (b) MAC-MORPHO  --  noticias em portugues do Brasil anotadas com POS
# =========================================================================
resumo_generico("(b) MAC-MORPHO", mac_morpho)

# O README esta gravado em UTF-8, mas mac_morpho.readme() o decodifica com a
# codificacao do corpus (latin-1) e gera acentos quebrados ("Núcleo...").
# Lemos o arquivo manualmente em UTF-8.
readme_mm = nltk.data.find("corpora/mac_morpho/README").open().read().decode("utf-8")
print("\n----- README (lido em UTF-8) -----")
print(readme_mm.strip())

# --- Atributos especificos ----------------------------------------------------
etiquetas = sorted({tag for _, tag in mac_morpho.tagged_words()})
gramaticais = [t for t in etiquetas if t[:1].isalpha()]
print(f"\npalavras={len(mac_morpho.words())}  frases={len(mac_morpho.tagged_sents())}"
      f"  etiquetas_distintas={len(etiquetas)}")
print("etiquetas gramaticais (amostra):", gramaticais[:22])
print(f"\nfrase etiquetada de exemplo (fileid = {mac_morpho.fileids()[0]}):")
print(mac_morpho.tagged_sents()[0][:12])


### (b) MAC-MORPHO  (reader: MacMorphoCorpusReader)
raiz     : C:\Users\andre\AppData\Roaming\nltk_data\corpora\mac_morpho
cod./arq.: latin-1 / 109 arquivo(s)  -> ['ag94ab12.txt', 'ag94ag02.txt', 'ag94de06.txt', 'ag94fe1.txt'] ...
metodos  : ['raw', 'words', 'sents', 'paras', 'tagged_words', 'tagged_sents']

----- README (lido em UTF-8) -----
MAC-MORPHO: Brazilian Portuguese news text with part-of-speech tags

1,167,183 words of journalistic texts extracted from ten sections of
the daily newspaper Folha de Sau Paulo, 1994.  (Version for training taggers.)

http://www.nilc.icmc.usp.br/lacioweb/

Distributed with permission of
Núcleo Interinstitucional de Lingüística Computacional (NILC),
Universidade de São Paulo (USP) in São Carlos,
Universidade Federal de São Carlos (UFSCar),
Universidade Estadual Paulista (UNESP) of Araraquara.

palavras=1170095  frases=51397  etiquetas_distintas=82
etiquetas gramaticais (amostra): ['ADJ', 'ADJ|+', 'ADJ|EST', 'ADV', 'ADV-KS', 'ADV-KS-REL', 'ADV|+', '

### (b) Mac-Morpho — características

<table>
<thead><tr><th>Aspecto</th><th>Descrição</th></tr></thead>
<tbody>
<tr><td><strong>O que é</strong></td><td>Corpus <strong>jornalístico do português do Brasil anotado morfossintaticamente</strong> (<em>part-of-speech tagging</em>). É a base de referência para treino/avaliação de etiquetadores POS em PT-BR.</td></tr>
<tr><td><strong>Origem do texto</strong></td><td>~1.167.183 palavras de textos jornalísticos extraídos de <strong>dez seções do jornal <em>Folha de S.Paulo</em>, ano de 1994</strong> (caderno de agropecuária, economia, esporte etc.).</td></tr>
<tr><td><strong>Estrutura / metadados</strong></td><td>Na distribuição do NLTK: <strong>109 arquivos</strong> <code>.txt</code> (versão para treino de <em>taggers</em>), ~1,17 milhão de tokens e ~51,4 mil frases. O <strong>nome de cada arquivo codifica a seção e a data</strong> do texto (ex.: <code>ag94ab12.txt</code> ≈ seção "agropecuária", abril de 1994).</td></tr>
<tr><td><strong>Anotação</strong></td><td>Uma etiqueta morfossintática por <em>token</em>. Conjunto de ~22 etiquetas básicas — <code>N</code> (substantivo), <code>V</code> (verbo), <code>ART</code>, <code>PREP</code>, <code>ADJ</code>, <code>ADV</code>, <code>NPROP</code> (nome próprio), <code>NUM</code>, <code>KC</code>/<code>KS</code> (conjunções), <code>PROPESS</code>, <code>PROADJ</code>, <code>PCP</code>, <code>CUR</code> (moeda) … — acrescidas de sufixos de contração/estrutura (<code>|+</code>, <code>|EST</code>, <code>|[</code>, <code>|]</code>), o que eleva o total a ~80 rótulos distintos no arquivo. Acesso via <code>words</code>, <code>sents</code>, <code>tagged_words</code>, <code>tagged_sents</code>.</td></tr>
<tr><td><strong>Anos dos textos</strong></td><td>1994.</td></tr>
<tr><td><strong>Produção / ano de produção</strong></td><td>Produzido pelo <strong>NILC – Núcleo Interinstitucional de Linguística Computacional</strong> (USP São Carlos), com UFSCar e UNESP/Araraquara, no âmbito do projeto <strong>Lácio-Web</strong> (início dos anos 2000). Distribuído com permissão do NILC.</td></tr>
<tr><td><strong>Codificação / licença</strong></td><td><code>latin-1</code>; uso acadêmico, distribuído com permissão.</td></tr>
<tr><td><strong>Uso típico</strong></td><td>Treino e avaliação de POS <em>taggers</em> para o português; estudos de frequência de classes gramaticais.</td></tr>
</tbody>
</table>


In [4]:
# =========================================================================
# (c) FLORESTA SINTACTICA  --  treebank do portugues (PT europeu + PT-BR)
# =========================================================================
import io
import contextlib

resumo_generico("(c) FLORESTA SINTACTICA", floresta)

print("\n----- README (cabecalho) -----")
print(floresta.readme()[:210].rstrip())
print("   [...]  chave das etiquetas: http://visl.sdu.dk/visl/pt/portsymbol.html")

# Percorrer TODA a Floresta imprime mensagens "Bad tree detected..." em
# sys.stderr: algumas frases do .ptb tem parenteses desbalanceados e o
# BracketParseCorpusReader as substitui por uma analise sintatica "plana".
# Nao e erro do nosso codigo; capturamos as mensagens so para manter a saida
# limpa e contamos quantas frases foram recuperadas.
_avisos = io.StringIO()
with contextlib.redirect_stderr(_avisos):
    n_palavras = len(floresta.words())
    frases = floresta.parsed_sents()
    n_frases = len(frases)
    etiquetas = sorted({tag for _, tag in floresta.tagged_words()})
    arvore = frases[0]
n_recup = _avisos.getvalue().count("Bad tree detected")

# --- Atributos especificos ----------------------------------------------------
print(f"\npalavras={n_palavras}  frases={n_frases}  etiquetas_distintas={len(etiquetas)}")
print(f"frases recuperadas com parse 'plano': {n_recup} ({n_recup / n_frases:.1%})")
print("etiquetas 'funcao+categoria' (amostra):",
      [t for t in etiquetas if any(c.isalpha() for c in t)][:8])
print("\nfrase 0 - tagged:", floresta.tagged_sents()[0])
print("frase 0 - arvore:", arvore)
arvore.pretty_print()


### (c) FLORESTA SINTACTICA  (reader: BracketParseCorpusReader)
raiz     : C:\Users\andre\AppData\Roaming\nltk_data\corpora\floresta
cod./arq.: ISO-8859-15 / 1 arquivo(s)  -> ['Floresta_7.4.ptb'] 
metodos  : ['raw', 'words', 'sents', 'tagged_words', 'tagged_sents', 'parsed_sents']

----- README (cabecalho) -----
Portuguese Treebank

Projecto Floresta Sinta(c)tica -- http://www.linguateca.pt/Floresta/
Version 7.4  Distributed with permission.

Penn Treebank format, available from http://linguateca.di.uminho.pt/FS/fs.htm
   [...]  chave das etiquetas: http://visl.sdu.dk/visl/pt/portsymbol.html

palavras=211852  frases=9266  etiquetas_distintas=269
frases recuperadas com parse 'plano': 131 (1.4%)
etiquetas 'funcao+categoria' (amostra): ['-H+prp', '>A+adj', '>A+adv', '>A+art', '>A+n', '>A+num', '>A+pp', '>A+pron-det']

frase 0 - tagged: [('Um', '>N+art'), ('revivalismo', 'H+n'), ('refrescante', 'N<+adj')]
frase 0 - arvore: (UTT+np (>N+art Um) (H+n revivalismo) (N<+adj refrescante))
       

### (c) Floresta Sintáctica — características

<table>
<thead><tr><th>Aspecto</th><th>Descrição</th></tr></thead>
<tbody>
<tr><td><strong>O que é</strong></td><td><em>Treebank</em> (banco de árvores sintáticas) do português — o subconjunto distribuído no NLTK corresponde ao <strong>"Bosque"</strong>, a parte revista manualmente do projeto, na <strong>versão 7.4</strong>, em <strong>formato Penn Treebank</strong>.</td></tr>
<tr><td><strong>Origem do texto</strong></td><td><strong>Projecto Floresta Sintá(c)tica</strong>, da Linguateca (<code>linguateca.pt/Floresta</code>), em parceria com o grupo VISL (Universidade do Sul da Dinamarca). Os textos são <strong>jornalísticos</strong>: excertos do jornal <em>Público</em> (português europeu) e da <em>Folha de S.Paulo</em> (português do Brasil), do fim dos anos 1990. As análises sintáticas foram geradas pelo <em>parser</em> <strong>PALAVRAS</strong> e revistas por linguistas.</td></tr>
<tr><td><strong>Estrutura / metadados</strong></td><td>No NLTK: <strong>1 arquivo</strong> (<code>Floresta_7.4.ptb</code>), ~211,8 mil palavras e <strong>~9.266 frases com análise sintática completa</strong>. O <code>README</code> traz a <strong>chave das etiquetas</strong> (padrão VISL, <code>visl.sdu.dk/visl/pt/portsymbol.html</code>).</td></tr>
<tr><td><strong>Anotação</strong></td><td>Análise sintática <strong>em constituintes</strong>, combinando, em cada nó/<em>token</em>, <strong>função sintática + categoria morfológica</strong>. As etiquetas de <em>token</em> têm a forma <code>função+categoria</code> — ex.: <code>&gt;N+art</code> (artigo, modificador pré-nominal), <code>H+n</code> (núcleo nominal), <code>N&lt;+adj</code> (adjetivo pós-nominal), <code>P+v-fin</code> (verbo finito). Há centenas de rótulos distintos (<code>SUBJ</code>, <code>ACC</code>, <code>PIV</code>, <code>ADVL</code>, <code>APP</code>, <code>UTT</code>, <code>NPHR</code>…). Acesso via <code>words</code>, <code>sents</code>, <code>tagged_words</code>, <code>tagged_sents</code> e <strong><code>parsed_sents</code></strong> (objetos <code>nltk.Tree</code>).</td></tr>
<tr><td><strong>Anos dos textos</strong></td><td>Notícias do final da década de 1990.</td></tr>
<tr><td><strong>Produção / ano de produção</strong></td><td>Projeto iniciado em <strong>2000</strong>; versão 7.4 distribuída com permissão. Mantido pela Linguateca (Portugal).</td></tr>
<tr><td><strong>Codificação / licença</strong></td><td><code>ISO-8859-15</code> (latin-9); distribuído com permissão para uso acadêmico.</td></tr>
<tr><td><strong>Uso típico</strong></td><td>Treino e avaliação de <em>parsers</em> e <em>chunkers</em>, extração de gramáticas, estudos de sintaxe do português.</td></tr>
</tbody>
</table>


### Síntese comparativa

<table>
<thead><tr><th></th><th>Machado</th><th>Mac-Morpho</th><th>Floresta Sintáctica</th></tr></thead>
<tbody>
<tr><td><strong>Tipo</strong></td><td>Literário, autor único</td><td>Jornalístico</td><td>Jornalístico</td></tr>
<tr><td><strong>Variedade</strong></td><td>PT-BR</td><td>PT-BR</td><td>PT-PT + PT-BR</td></tr>
<tr><td><strong>Fonte</strong></td><td>Obras Completas (Nova Aguilar, 1994)</td><td><em>Folha de S.Paulo</em></td><td><em>Público</em> + <em>Folha de S.Paulo</em></td></tr>
<tr><td><strong>Anos dos textos</strong></td><td>1854–1939</td><td>1994</td><td>fim dos anos 1990</td></tr>
<tr><td><strong>Anotação</strong></td><td>nenhuma (<em>plaintext</em>)</td><td>POS (morfossintática)</td><td>sintática (constituintes + funções)</td></tr>
<tr><td><strong>Arquivos (NLTK)</strong></td><td>246</td><td>109</td><td>1</td></tr>
<tr><td><strong>Tamanho aprox.</strong></td><td>~246 obras</td><td>~1,17 M palavras / ~51 k frases</td><td>~212 k palavras / ~9,3 k frases</td></tr>
<tr><td><strong>Reader NLTK</strong></td><td><code>CategorizedPlaintextCorpusReader</code></td><td><code>MacMorphoCorpusReader</code></td><td><code>BracketParseCorpusReader</code></td></tr>
<tr><td><strong>Método-chave</strong></td><td><code>words</code>, <code>sents</code>, <code>paras</code>, <code>categories</code></td><td><code>tagged_words</code>, <code>tagged_sents</code></td><td><code>parsed_sents</code> (<code>nltk.Tree</code>)</td></tr>
</tbody>
</table>

**Referências:**

- CARVALHO, Fabrício Galende Marques de. *Introdução ao Processamento de Linguagem Natural*.
- Repositório da disciplina: <https://github.com/fabriciogmc/natural_language_processing>
- Documentação do NLTK — *Accessing Text Corpora and Lexical Resources* (`https://www.nltk.org/howto/corpus.html`) e NLTK Book, cap. 2.
- Portal Machado de Assis — Obra Completa (`http://machado.mec.gov.br`).
- NILC / projeto Lácio-Web (`http://www.nilc.icmc.usp.br/lacioweb/`).
- Projecto Floresta Sintáctica / Linguateca (`https://www.linguateca.pt/Floresta/`).
- Arquivos `README` de cada corpus (impressos nas células acima).
